### PR TITLE
Allow logs to be mutated by LogProcessor

### LINKED ISSUE #4550: How to modify log data in the log pipeline?
There's AutoConfigurationCustomizer.addLogExporterCustomizer which gives a nice hook to modify log data, but it occurs after the BatchLogProcessor, and so doesn't have access to the Context.

LogProcessors run before the BatchLogProcessors, but are not able to modify log data (which makes them unlike SpanProcessors, but also makes them much simpler from concurrency/performance perspective).

Short of making LogProcessors similar to SpanProcessors and allowing them to modify the logs, wdyt of adding
AutoConfigurationCustomizer.addBatchLogProcessorCustomizer to allow modifying log data before it goes to the BatchLogProcessor which you still have access to the Context?

### PR DESCRIPTION
**Pull request by jack-berg**
Implements spec [PR#2681](https://github.com/open-telemetry/opentelemetry-specification/pull/2681).
Resolves #4550.

Summary:
- Rename `LogBuilder` to `LogRecordBuilder` to better reflect the data model and the naming we can expect to see in the log appender / event API. 
- Change `LogProcessor#emit` to `onEmit` to better match trace SDK. 
- Introduce `ReadWriteLogRecord` interface, which is passed to `LogProcessor#onEmit` instead of `LogData`. This is roughly aligned with tracing. However, in tracing `ReadWriteSpan` extends `Span` and `ReadableSpan`. This separation of `Span` for updating and `ReadableSpan` for reading isn't needed in the log SDK because there isn't a `LogProcessor` `onStart` and `onEnd` which have different read / write semantics. 
- `ReadWriteLogRecord` has a `toLogData()` method, which is called by the simple and batch log processor to convert the data to an immutable representation before calling `LogExporter#export(Collection<LogData>)`. 
- Add single attribute setters (`setAttribute(AttributeKey<T>, T)`), rename multi attribute setter to `setAllAttributes(Attributes)` to mirror tracing

### CODE REVIEW

**Comment by jack-berg** [comment]
Hey @jkwatson - wondering what your thoughts are about trying to get this in before the 1.17.0 release. It technically depends on this [spec PR](https://github.com/open-telemetry/opentelemetry-specification/pull/2681), but I think its extremely likely that the spec goes in this direction and that its current state of not allowing mutable logs is accidental. Additionally, both the spec and the sdk are experimental so I think its ok for us to experiment (within reason!).

**Review by jkwatson — APPROVED** [review]
I'm fine with this. It'll be interesting to see how it ends up differing from the final spec, but we'll update as appropriate before stabilization.

File: sdk/logs/src/main/java/io/opentelemetry/sdk/logs/ReadWriteLogRecord.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+/** A log record that can be read from and written to. */
+public interface ReadWriteLogRecord {
+
+  /**
+   * Sets an attribute on the log record. If the log record previously contained a mapping for the
+   * key, the old value is replaced by the specified value.
+   *
+   * <p>Note: the behavior of null values is undefined, and hence strongly discouraged.
+   */
+  <T> ReadWriteLogRecord setAttribute(AttributeKey<T> key, T value);
```
Comment (jack-berg) [inline sdk/logs/src/main/java/io/opentelemetry/sdk/logs/ReadWriteLogRecord.java:20]: @trask for now the only thing you can mutate is log attributes. I plan on adding getters / setters on here in the future, but wondering if this is sufficient for your short term requirements.

Comment (trask) [reply sdk/logs/src/main/java/io/opentelemetry/sdk/logs/ReadWriteLogRecord.java:20]: yes, this is exactly what I need for now 👍

**Pull request merged**